code/modules/projectiles/ammunition/smartmag.dm

```diff
@@ -0,0 +1,168 @@
+	if(!attached_cell)
+		user << "<span class='warning'>\The [src] does not appear to have a power source installed.</span>"
+
+/obj/item/ammo_magazine/smart/update_icon()
+	if(attached_cell)
+		icon_state = "smartmag-filled"
+	else if(!(attached_cell))
+		icon_state = "smartmag-empty"
+
+/obj/item/ammo_magazine/smart/proc/chargereduction()
+	if(attached_cell)
+		if(attached_cell.checked_use(production_cost))
+			return 1
+		else
+			return 0
+
+	return null
+
+/obj/item/ammo_magazine/smart/proc/set_production_cost(var/obj/item/ammo_casing/A)
+	var/list/matters = ammo_repository.get_materials_from_object(A)
+	var/tempcost
+	for(var/key in matters)
+		var/value = matters[key]
+		tempcost += value * production_modifier
+	production_cost = tempcost
+
+/obj/item/ammo_magazine/smart/attack_self(mob/user)
+	if(emagged)
+		return ..()
+	else
+		user << "<span class='notice'>\The [src] is not designed to be unloaded.</span>"
+		return
+
+/obj/item/ammo_magazine/smart/emag_act(var/remaining_charges, var/mob/user)
+	if(!emagged)
+		user << "<span class='notice'>You overload \the [src]'s security measures causing widespread destabilisation. It is likely you could empty \the [src] now.</span>"
+		emagged = 1
+		return 1
+
+/obj/item/ammo_magazine/smart/emp_act(severity)
+	..()
+	if(attached_cell)
+		attached_cell.emp_act(severity)
+
+/obj/item/ammo_magazine/smart/attackby(var/obj/item/I as obj, mob/user)
+	if(istype(I, /obj/item/weapon/cell))
+		if(attached_cell)
+			user << "<span class='notice'>\The [src] already has a [attached_cell.name] attached.</span>"
+			return
+		else
+			user << "You begin inserting \the [I] into \the [src]."
+			if(do_after(user, 25))
+				user.drop_item()
+				I.loc = src
+				attached_cell = I
+				user.visible_message("[user] installs a cell in \the [src].", "You install \the [I] into \the [src].")
+				update_icon()
+				return
+
+	else if(istype(I, /obj/item/weapon/screwdriver))
+		if(attached_cell)
+			user << "You begin removing \the [attached_cell] from \the [src]."
+			if(do_after(user, 25))
+				attached_cell.update_icon()
+				attached_cell.loc = get_turf(src.loc)
+				attached_cell = null
+				user.visible_message("[user] removes a cell from \the [src].", "You remove \the [attached_cell] from \the [src].")
+				update_icon()
+				return
+
+	var/obj/item/ammo_magazine/A
+	var/obj/item/weapon/gun/projectile/G
+	var/obj/item/ammo_casing/C
+
+	if(istype(I, /obj/item/ammo_magazine))
+		A = I
+	else if(istype(I, /obj/item/weapon/gun/projectile))
+		G = I
+	else if(istype(I, /obj/item/ammo_casing))
+		C = I
+	else
+		return
+
+	//if(G && caliber == G.caliber && ammo_type == G.ammo_type || A && caliber == A.caliber && ammo_type == A.ammo_type)
+	if(G && G.ammo_magazine && caliber == G.caliber && ammo_type == G.ammo_magazine.type || A && caliber == A.caliber && ammo_type == A.ammo_type || C && caliber == C.caliber && ammo_type == C.projectile_type)
```

Comment:
This can be shortened a bit by doing instead
```
	var/new_caliber = caliber
	var/new_ammo_type = ammo_type
	
	if(istype(I, /obj/item/ammo_magazine))
		var/obj/item/ammo_magazine/A = I
		new_caliber = A.caliber
		new_ammo_type = A.ammo_type
		
	else if(istype(I, /obj/item/weapon/gun/projectile))
		var/obj/item/weapon/gun/projectile/G = I
		if(G.ammo_magazine)
			new_caliber = G.ammo_magazine.caliber
		new_ammo_type = G.ammo_magazine.ammo_type
		
	else if(istype(I, /obj/item/ammo_casing))
		var/obj/item/ammo_casing/C = I
		new_caliber = C.caliber
		new_ammo_type = C.projectile_type
	else
		return

	if(caliber == new_caliber && ammo_type == new_ammo_type)
		..()
	else
		caliber = new_caliber
		ammo_type = new_ammo_type
		set_production_cost(new_ammo_type)
	
		user << "<span class='notice'>You scan \the [I] with \the [src], copying \the [I]'s caliber and ammunition type.</span>"
		stored_ammo.Cut()
```